numpyro/infer/ensemble.py

```diff
@@ -0,0 +1,724 @@
+                                                                                  jnp.array(0)
+                                                                                  ),  
+                                                            lambda _, __: (mu, 
+                                                                           n_expansions,
+                                                                           n_contractions
+                                                                           ),
+                                                            safe_n_expansions, n_contractions)
+   
+        return proposal, ESSState(itr, n_expansions, n_contractions, mu, rng_key)
+
+
+    @staticmethod
+    def RandomMove():
+        """
+        The `Karamanis & Beutler (2020) <https://arxiv.org/abs/2002.06212>`_ "Random Move" with parallelization.
+        When this move is used the walkers move along random directions. There is no communication between the
+        walkers and this Move corresponds to the vanilla Slice Sampling method. This Move should be used for
+        debugging purposes only.
+        """
+        def random_move(rng_key, inactive, mu):
+            directions = dist.Normal(loc=0, scale=1).sample(
+                rng_key, sample_shape=inactive.shape
+            )
+            directions /= jnp.linalg.norm(directions, axis=0)
+
+            return 2.0 * mu * directions
+        return random_move
+
+    @staticmethod
+    def KDEMove(bw_method=None):
+        """
+        The `Karamanis & Beutler (2020) <https://arxiv.org/abs/2002.06212>`_ "KDE Move" with parallelization.
+        When this Move is used the distribution of the walkers of the complementary ensemble is traced using
+        a Gaussian Kernel Density Estimation methods. The walkers then move along random direction vectos
+        sampled from this distribution.
+        """
+        def kde_move(rng_key, inactive, mu):
+            n_active_chains, n_params = inactive.shape
+
+            kde = gaussian_kde(inactive.T, bw_method=bw_method)
+
+            vectors = kde.resample(rng_key, (2 * n_active_chains,)).T
+            directions = vectors[:n_active_chains] - vectors[n_active_chains:]
+
+            return 2.0 * mu * directions
+        return kde_move
+
+    @staticmethod
+    def GaussianMove():
+        """
+        The `Karamanis & Beutler (2020) <https://arxiv.org/abs/2002.06212>`_ "Gaussian Move" with parallelization.
+        When this Move is used the walkers move along directions defined by random vectors sampled from the Gaussian
+        approximation of the walkers of the complementary ensemble.
+        """
+        def gaussian_move(rng_key, inactive, mu):
+            n_active_chains, n_params = inactive.shape
+            cov = jnp.cov(inactive, rowvar=False)
+
+            return (
+                2.0
+                * mu
+                * dist.MultivariateNormal(0, cov).sample(
```

Comment:
haven't looked through your code in detail but i believe you can get the MVN sample much more cheaply by "coloring" (i.e. multiplying) a vector of `N(0,1)` normal noise by `inactive`. if you use `MultivariateNormal` like this it will compute a cholesky decomposition at cubic cost

Reply:
These don't seem to be sampling from the same distribution - am I misunderstanding what you meant?

```python
import numpy as np
import matplotlib.pyplot as plt

n_obs = 1_000
n_dim = 2

particles = np.random.exponential(10, (n_obs, n_dim))
cov = np.cov(particles, rowvar=False)

cholesky_samp = np.random.multivariate_normal(np.mean(particles, axis=0),
                                              cov=cov,
                                              size=(n_obs,))


cheaper_samp = particles * np.random.standard_normal((n_obs, n_dim))

plt.scatter(cholesky_samp[:, 0], cholesky_samp[:, 1], alpha=.3)
plt.scatter(cheaper_samp[:, 0], cheaper_samp[:, 1], alpha=.3)
```
![8e676227-5ac4-4d94-adfe-d782e19fea93](https://github.com/pyro-ppl/numpyro/assets/96739930/426c985a-8087-4779-86dc-0d066ac8467c)

Reply:
It seems like avoiding the cholesky decomposition is only faster in high dimensions (for low enough n_obs), but gradient-free methods tend to fail there anyway. Do you think it's worth special casing it?

![85573bf1-2da3-4d8e-b9e8-9abde8e1d913](https://github.com/pyro-ppl/numpyro/assets/96739930/97892547-a062-4968-91f3-2346eda35e3f)
![816ed6d7-f5dc-4ff8-a2ec-2fd6a6e3f6cf](https://github.com/pyro-ppl/numpyro/assets/96739930/a227c768-caaf-4c4f-b278-7dac931f7dbc)

```python
#%%
from functools import partial
import jax
import jax.numpy as jnp
import jax.random as jrandom
import numpyro.distributions as dist
import matplotlib.pyplot as plt

@partial(jax.jit, static_argnums=2)
def no_chol_mvn(rng_key, data, n_samps):
    n_obs, n_dim = data.shape
    eps = jrandom.normal(rng_key, (n_samps, n_obs))

    return eps @ (data - jnp.mean(data, axis=0)) / jnp.sqrt(n_obs)

@partial(jax.jit, static_argnums=2)
def chol_mvn(rng_key, data, n_samps):
    cov = jnp.cov(data, rowvar=False, bias=True)

    return dist.MultivariateNormal(0, cov).sample(rng_key, (n_samps,))

n_obs = n_samps = jnp.geomspace(10, 10_000, num=20, dtype=jnp.int32) 
n_dim = 20

no_chol_timings = []
chol_timings = []

for obs, samps in zip(n_obs, n_samps):
    data = jrandom.exponential(jrandom.PRNGKey(0), (obs, n_dim))

    no_chol_res = %timeit -o no_chol_mvn(jrandom.PRNGKey(1), data, samps.item()).block_until_ready()
    chol_res = %timeit -o chol_mvn(jrandom.PRNGKey(1), data, samps.item()).block_until_ready()

    no_chol_timings.append(no_chol_res)
    chol_timings.append(chol_res)    

no_chol_timings = [sum(res.timings) / len(res.timings) for res in no_chol_timings]
chol_timings = [sum(res.timings) / len(res.timings) for res in chol_timings]

#%%
plt.scatter(n_obs[:15], no_chol_timings[:15], label='no_chol')
plt.scatter(n_obs[:15], chol_timings[:15], label='chol')
plt.title(f'time vs. n_obs/n_samps, n_dim={n_dim}')
plt.legend();
```

Reply:
yeah if you only anticipate running this algorithm in up to ~50-100 dimensions it's probably not worth it but you can at least leave a comment in the code